clang/lib/Analysis/UnsafeBufferUsage.cpp

```diff
@@ -353,23 +355,90 @@ isInUnspecifiedUntypedContext(internal::Matcher<Stmt> InnerMatcher) {
   return stmt(anyOf(CompStmt, IfStmtThen, IfStmtElse));
 }
 
+// Returns true iff integer E1 is equivalent to integer E2.
+//
+// For now we only support such expressions:
+//    expr := DRE | const-value | expr BO expr
+//    BO   := '*' | '+'
+//
+// FIXME: We can reuse the expression comparator of the interop analysis after
```

Comment:
@ziqingluo-90 Is the expression comparator of the interop analysis the same or different from what is done here?

 I believe this expression comparison can result in exponential work in the size of an expression..  AreEquaIntegralBinaryOperator makes 2 calls to AreEqualIntegers per subexpression.  AreEqualIntegers can call back into EqualIntegeralBinaryOperator on subexpressions of its argument.

If the plan is to eventually replace this code with the expression comparator from interop analysis, then this is not worth fixing now.  However, if the expression comparator has the same issue or that is not the plan, then this should be fixed.   It can be fixed be bounding the amount of work that is allowed, for example, by limiting the recursive depth that is allowed.

Reply:
The comparator used in the interop analysis is more general. But it does not know anything about commutativity.  So the complexity issue will still be there if we use the same algorithm as here.  I plan to apply commutativity to only leaf nodes to fix the issue.

For associativity---yes, it's dangerous!  So I do not plan to support it.  I think my algorithm here does not introduce associativity otherwise I'd like to fix it right away.